clash-cores/src/Clash/Cores/Xilinx/Ila/Internal.hs

```diff
@@ -0,0 +1,360 @@
+  -- ^ Probe names. Clash will error if it cannot generate names passed here.
+  , probeTypes :: Either ProbeType (Vec n ProbeType)
+  -- ^ Purpose of probe. If 'Left', all probes will be set to the same type. If
+  -- 'Right', each probe type can be set individually. Also see 'ProbeType'.
+  , captureControl :: Bool
+  -- ^ Whether probes marked 'Trigger' or 'DataAndTrigger' can be used to control
+  -- data capture. That is, a trigger marks the start of data collection, while
+  -- capture control marks when to sample.
+  }
+  deriving (Show, Lift)
+
+-- XXX: I'd move this 'deriveTermLiteral' up, but Template Haskell complains..
+deriveTermLiteral ''ProbeType
+instance KnownNat n => TermLiteral (IlaConfig n) where
+  termToData = $(deriveTermToData ''IlaConfig)
+
+probeTypesVec :: KnownNat n => IlaConfig n -> Vec n ProbeType
+probeTypesVec = either C.repeat id . probeTypes
+
+comparatorsVec :: KnownNat n => IlaConfig n -> Vec n Int
+comparatorsVec = either C.repeat id . comparators
+
+areSame :: Eq a => [a] -> Maybe a
+areSame = \case
+  c:cs -> foldl go (Just c) cs
+  [] -> Nothing
+ where
+  go Nothing _ = Nothing
+  go (Just c0) c1
+    | c0 == c1 = Just c0
+    | otherwise = Nothing
+
+ilaBBF :: HasCallStack => BlackBoxFunction
+ilaBBF _isD _primName args _resTys
+  | _:_:config:_ <- lefts args
+  , _:_:LitTy (NumTy n):_ <- rights args
+  , Just (SomeNat (Proxy :: Proxy n)) <- someNatVal n
+  = case termToDataError @(IlaConfig n) config of
+      Left s -> error ("ilaBBF, bad config:\n" <> s)
+      Right c -> pure $ Right (bbMeta c, bb c)
+  | otherwise = error $ "ilaBBF, bad args:\n" <> ppShow args
+ where
+   bbMeta :: KnownNat n => IlaConfig n -> BlackBoxMeta
+   bbMeta config = emptyBlackBoxMeta
+     { bbKind = TDecl
+     , bbRenderVoid = RenderVoid
+     , bbIncludes =
+         [ ( ("ila", "clash.tcl")
+           , BBFunction (show 'ilaTclTF) 0 (ilaTclTF config)
+           )
+         ]
+     }
+
+   bb :: KnownNat n => IlaConfig n -> BlackBox
+   bb config = BBFunction (show 'ilaTF) 0 (ilaTF config)
+
+usedArguments :: [Int]
+usedArguments = (ilaConfig : clock : inputProbes)
+ where
+  (    _knownDomain
+    :< _ilaConstraint
+    :< ilaConfig
+    :< clock
+    :< (Infinite.take 8096 -> inputProbes)
+    ) = ((0::Int)...) -- This function is polyvariadic so in theory it supports an
+                      -- unlimited number of arguments. To prevent evaluation loops
+                      -- when forcing this argument to NF we limit it to a modest
+                      -- 8096 input ports.
+
+ilaTF :: (HasCallStack, KnownNat n) => IlaConfig n -> TemplateFunction
+ilaTF config =
+  TemplateFunction
+    usedArguments
+    -- 'validateIlaProbeBCC' already produces string describing
+    -- failing checks, but 'TemplateFunction' cannot handle this
+    -- yet. This is prepared to get updated easily as soon as the
+    -- feature gets implemented in clash-lib.
+    (maybe True error . validateIlaProbeBBC)
```

Comment:
Nice, although my suggestion for handling errors is that `TemplateFunction` should be changed to

```haskell
data TemplateFunction where
  TemplateFunction
    :: [Int]
    -- ^ Used arguments
    -> (forall s . Backend s => BlackBoxContext -> ExceptT Text (Lazy.State s Doc))
    -- ^ Render function
    -> TemplateFunction
```
so we can just handle all errors in the render function instead of some here and some there.